File: 백준/Silver/3273. 두 수의 합/두 수의 합.cc
#include <iostream>
using namespace std;

int check[2000005];
int ans;
int main() {
	int n,x;
	cin >> n;
	int arr[n];

	for(int i = 0; i < n; i++) {
		cin >> arr[i];
		check[arr[i]]++;
	}
	
	cin >> x;

	for(int i = 0; i < n; i++) {
		if(x-arr[i] == arr[i]) continue;
		if(x-arr[i] <= 0) continue;
		if(check[x-arr[i]]) {
			ans++;
		}
	}
	cout << ans / 2;
	return 0;
}

File: 백준/Silver/3273. 두 수의 합/두 수의 합.java
import java.io.*;
import java.util.*;

class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		StringTokenizer st = new StringTokenizer(br.readLine());
		int[] arr = new int[n];
		for(int i = 0; i < n; i++){
			arr[i] = Integer.parseInt(st.nextToken());
		}
		int x = Integer.parseInt(br.readLine());
		
		Arrays.sort(arr);
		int left = 0, right = arr.length - 1,ans = 0;
		while(left < right){
			if(arr[left] + arr[right] < x) left++;
			else if(arr[left] + arr[right] > x) right--;
			else{
				ans++;
				left++;
				right--;
			}
		}
		System.out.print(ans);
	}
}